openwisp_users/api/mixins.py

```diff
@@ -125,26 +125,29 @@ def _user_attr(self):
         raise NotImplementedError()
 
     def filter_fields(self):
-        user = self.context['request'].user
-        if user.is_superuser:
-            return
-        organization_filter = getattr(user, self._user_attr)
-        for field in self.fields:
-            if field == 'organization':
-                self.fields[field].allow_null = False
-                self.fields[field].queryset = self.fields[field].queryset.filter(
-                    pk__in=organization_filter
-                )
-                continue
-            try:
-                conditions = Q(**{self.organization_lookup: organization_filter})
-                if self.include_shared:
-                    conditions |= Q(organization__isnull=True)
-                self.fields[field].queryset = self.fields[field].queryset.filter(
-                    conditions
-                )
-            except AttributeError:
-                pass
+        try:
```

Comment:
never wrap an entire block in a try statement, only wrap the line which you expect can fail, otherwise we risk catching other errors as well but we won't understand what's going on and will be hard to debug.

Reply:
@nemesisdesign the error is raised due to this line:- 
`user = self.context['request'].user`
I tried wrapping only this line in `try` statement but then other variables like `organization_filter` depends on it with which entire logic is dependent, it raises variable used before assignment error. 

I tried assigning None to it, but it didn't work.

Reply:
Indeed, too many folks add all the code in the try block, we should avoid this.

> tried wrapping only this line in try statement but then other variables like organization_filter depends on it with which entire logic is dependent

Why not `return` in the catch block when getting user is failed?
Wouldn't that solve the problem?

Reply:
Done @atb00ker

Reply:
@ManishShah120 why would the request be missing from the serializer context?
Is there a legitimate reason for this to happen?
I think we should fail the request if `KeyError` is raised, because it means the application is misconfigured.
The serializer shall always have the request passed to the serializer context during initialization or otherwise it won't work.

Don't we risk a security issue if the return statement is hit?
I suggest simply raising [APIException](https://github.com/encode/django-rest-framework/blob/master/rest_framework/exceptions.py#L96-L130) if the except block is reached.

Reply:
> @ManishShah120 why would the request be missing from the serializer context?

@nemesisdesign when we use nested serializer, then the `self.context` dict is `empty` when the server is started so it cannot get the `request` object from the context hence `KeyError` is raised. But once the server has started successfully then when send a request to the endpoint containing the nested serializer is sent, then the `self.context` dict contains all the required objects.

We are required to so because the server doesn't start, due to this KeyError being raised.

I tried raising `APIException` but again the server doesn't start raising this error:-
```
  File "/home/manish/Videos/openwisp_users_remote/openwisp-users/openwisp_users/api/mixins.py", line 132, in filter_fields
    raise APIException
rest_framework.exceptions.APIException: A server error occurred.
```

Reply:
let me know how to replicate this issue and I'll test it

Reply:
@nemesisdesign To replicate this issue simply fetch the changes I have pushed to this branch https://github.com/openwisp/openwisp-firmware-upgrader/tree/openwisp-users-api-mixins ,and update openwisp-users to latest master. Then start the server.

Ps: This error is raised when we use `FilterSerializerByOrgManaged` in a serializer class which contains nested serilaizer field.

Or you can undo the changes and see the tests.